LoginClient/PipeClient.cs

```diff
@@ -0,0 +1,368 @@
+
+                            bytesRead += numBytes;
+
+                        } while (bytesRead < totalSize);
+                    }
+                    catch
+                    {
+                        break;
+                    }
+
+                    // Client has disconnected.
+                    if (bytesRead == 0)
+                    {
+                        break;
+                    }
+
+                    // Call message recieved event.
+                    if(MessageRecieved != null)
+                    {
+                        MessageRecieved(ms.ToArray());
+                        Console.Out.WriteLine(ms.ToArray());
+                        Console.Out.WriteLine(Thread.CurrentThread.ManagedThreadId + " in read()");
+
+                        if (!IsLoggedIn())
+                        {
+                            Console.Out.WriteLine();
+                            Console.Out.WriteLine("Begin validation.");
+                            ValidationResult(ms.ToArray());
+                            Console.Out.WriteLine("Validation has ended");
+                            Console.Out.WriteLine();
+                        }
+                    }
+                }
+            }
+
+            // If disconnected then the disconnection was caused by the server
+            // being terminated.
+            if (IsConnected())
+            {
+                // Clean up the resources.
+                stream.Close();
+                handle.Close();
+
+                stream = null;
+                handle = null;
+
+                // Client no longer conected to the server.
+                IsLoggedIn(false);
+                IsConnected(false);
+                SetPipeName(null);
+
+                if (ServerDisconnected != null)
+                {
+                    ServerDisconnected();
+                }
+            }
+
+        }
+
+
+        /// <summary>
+        /// Sends the message to the server.
+        /// </summary>
+        /// <param name="message"></param>
+        /// <returns></returns>
+        public void SendMessage(byte[] message)
+        {
+            // Check what the two stream.write lines do.
+            ASCIIEncoding encoder = new ASCIIEncoding();
+            string str = encoder.GetString(message, 0, message.Length);
+            Console.Out.WriteLine();
+            Console.Out.WriteLine("The thread Sending the message is thread: " + Thread.CurrentThread.ManagedThreadId);
+            Console.Out.WriteLine("The message sent to the Send Message method is: \n" + str);
+            Console.Out.WriteLine();
+            Console.Out.WriteLine(BitConverter.GetBytes(message.Length));
+            Console.Out.WriteLine();
+
+            try
+            {
+                // Write the entire stream length.
+                stream.Write(BitConverter.GetBytes(message.Length), 0, 4);
+
+
+                stream.Write(message, 0, message.Length);
```

Comment:
Confused.  So might be nothing.

Your writting to the stream twice?  the message first 4 bytes, and then the message in totality?
weird. but if thats what your meant to do, ok.